src/Symfony/Component/Mailer/Bridge/Amazon/Transport/SesTransportFactory.php

```diff
@@ -21,23 +24,41 @@
  */
 final class SesTransportFactory extends AbstractTransportFactory
 {
+    private $credentialProvider;
+
+    public function __construct(EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null)
+    {
+        parent::__construct($dispatcher, $client, $logger);
+        
+        $this->credentialProvider = new InstanceCredentialProvider($client);
+    }
+
     public function create(Dsn $dsn): TransportInterface
     {
         $scheme = $dsn->getScheme();
-        $user = $this->getUser($dsn);
-        $password = $this->getPassword($dsn);
+        try {
+            $credential = new UsernamePasswordCredential($this->getUser($dsn), $this->getPassword($dsn));
+        } catch (IncompleteDsnException $e) {
+            $role = $dsn->getOption('role');
+
+            if (null === $role) {
+                throw new IncompleteDsnException('User and password nor role is not set.');
+            }
+
+            $credential = $this->credentialProvider->getCredential($role);
+        }
         $region = $dsn->getOption('region');
 
         if ('api' === $scheme) {
-            return new SesApiTransport($user, $password, $region, $this->client, $this->dispatcher, $this->logger);
+            return new SesApiTransport($credential, $region, $this->client, $this->dispatcher, $this->logger);
         }
 
         if ('http' === $scheme) {
-            return new SesHttpTransport($user, $password, $region, $this->client, $this->dispatcher, $this->logger);
+            return new SesHttpTransport($credential, $region, $this->client, $this->dispatcher, $this->logger);
         }
 
         if ('smtp' === $scheme || 'smtps' === $scheme) {
-            return new SesSmtpTransport($user, $password, $region, $this->dispatcher, $this->logger);
+            return new SesSmtpTransport($credential, $region, $this->dispatcher, $this->logger);
```

Comment:
what if an ApiTokenCredential was instantiated and we reach this point ?

Reply:
@stof I think the supported credential type(s) should be checked by the Transport instance.

Reply:
@stof SesSmtpTransport only works with UsernamePasswordCredential (as you can see at the first parameter of the constructor). DX would be better if we check the class and throw an exception with a better message than a native type exception.